### ISSUE TITLE
Disallow null for optional strings

### ISSUE THREAD

**Issue by tmacht** [domaindrivendev/Swashbuckle.AspNetCore#1218]
When updating from version 4.0.1 to 5.0.0-rc2 we encounter the following problem: String properties that are NOT marked as Required are suddenly created with "Disallow null". I think it is a big problem that is not possible to have nullable string properties or is there another way to achieve this in v5?

**Comment by domaindrivendev** [comment]
Please provide exact repro steps

**Comment by tmacht** [comment]
I have attached an example solution, see GetCorrespondenceAddress in the ValuesController:
[OptionalStringsNotNullable.zip](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/files/3476019/OptionalStringsNotNullable.zip)

The CorrespondenceAddress has one required string property (Street) and two optional properties ( HouseNumber, HouseNumberExtension):

```
public class CorrespondenceAddress
{
    [Required]
    public string Street { get; set; }

    public int? HouseNumber { get; set; }

    public string HouseNumberExtension { get; set; }
}
```

In the swagger.json, the following is created:

```
"schemas": {
      "CorrespondenceAddress": {
        "required": [
          "street"
        ],
        "type": "object",
        "properties": {
          "street": {
            "type": "string"
          },
          "houseNumber": {
            "type": "integer",
            "format": "int32",
            "nullable": true
          },
          "houseNumberExtension": {
            "type": "string"
          }
        },
        "additionalProperties": false
      }
```
The houseNumber is nullable as expected, but the houseNumberExtension is not.

We then use nswag to create a client (the .nswag file and the resulting generated code are also included in the example project).

The CorrespondenceAddress class in the client is created as follows:
```
[System.CodeDom.Compiler.GeneratedCode("NJsonSchema", "10.0.21.0 (Newtonsoft.Json v11.0.0.0)")]
public partial class CorrespondenceAddress 
{
    [Newtonsoft.Json.JsonProperty("street", Required = Newtonsoft.Json.Required.Always)]
    [System.ComponentModel.DataAnnotations.Required(AllowEmptyStrings = true)]
    public string Street { get; set; }

    [Newtonsoft.Json.JsonProperty("houseNumber", Required = Newtonsoft.Json.Required.Default, NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore)]
    public int? HouseNumber { get; set; }

    [Newtonsoft.Json.JsonProperty("houseNumberExtension", Required = Newtonsoft.Json.Required.DisallowNull, NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore)]
    public string HouseNumberExtension { get; set; }
}
```

However, I would expect that the HouseNumberExtension has Required.Default, just as the HouseNumber.

**Comment by domaindrivendev** [comment]
This should be fixed by 0f7f6e3fa47c76bd418b2ea5e9d1683e54562e84

With that said, it's worth noting that, with the OpenAPI 3 (OA3) specification, there's a difference between required and non-nullable. A required property MUST be provided in the payload but it's value MAY be null, as indicated by the `nullable` property in the spec. So, with the latest change, the following schema definition will be generated:


```json
"schemas": {
      "CorrespondenceAddress": {
        "required": [
          "street"
        ],
        "type": "object",
        "properties": {
          "street": {
            "type": "string",
            "nullable": true
          },
          "houseNumber": {
            "type": "integer",
            "format": "int32",
            "nullable": true
          },
          "houseNumberExtension": {
            "type": "string",
            "nullable": true
          }
        },
        "additionalProperties": false
      }
```

@tmacht please let me know if this makes sense?

**Comment by mkontula** [comment]
@domaindrivendev, is 0f7f6e3 already in 5.0.0-rc3-preview-0954 ? I createad equal test case but "nullable" was not generated to json definition.

**Comment by domaindrivendev** [comment]
@mkontula - perhaps you could confirm but I think now the issue is a little different? With `preview-0954` `nullable: true` is correctly being emitted for the string properties but not for the nullable int? At least that's what I'm seeing when I try to repro.

Either way, it's an issue that needs to be addressed before the official Nuget goes out. Re-opening and assigning to the v5 milestone.

**Comment by mkontula** [comment]
I can confirm, preview-0954, nullable: true is being emitted to string properties. 
Testing with the class below, nullable int and datetime value types did not get the `nullable: true`
```
public class ValueModelDto  
{
    [Required]
    public int Id { get; set; }
    [Required]
    public string RequiredString { get; set; }
    public string OptionalString { get; set; }
    [Required]
    public int RequiredInt { get; set; }
    public int? OptionalInt { get; set; }
    [Required]
    public DateTime RequiredDate { get; set; }
    public DateTime? OptionalDate { get; set; }
}
```

**Comment by domaindrivendev** [comment]
Should be fixed by 24d166d804b677d00747428c2561744e682d10e7 and available for test in the latest preview nuget on myget.org (`5.0.0-rc3-preview-0959` at time of writing)

**Comment by mkontula** [comment]
Nice! Thanks for the fix. Works like a charm!
